src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java

```diff
@@ -76,34 +79,56 @@ public void run() {
-                        if (igoCmoRequestFilter && !request.isCmoRequest()) {
-                            requestStatusLogger.logRequestStatus(request.getRequestJson(),
-                                    RequestStatusLogger.StatusType.CMO_REQUEST_FILTER_SKIPPED_REQUEST);
-                            continue;
-                        }
+                    String requestJson = cmoLabelGeneratorQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (requestJson != null && requestJson.length() > 10) {
                         // skip request if there are no samples to generate cmo labels for
-                        if (request.getSamples().isEmpty()) {
-                            requestStatusLogger.logRequestStatus(request.getRequestJson(),
+                        List<SampleMetadata> samples = getSamplesFromRequestJson(requestJson);
+                        if (samples.isEmpty()) {
+                            requestStatusLogger.logRequestStatus(requestJson,
                                     RequestStatusLogger.StatusType.REQUEST_WITH_MISSING_SAMPLES);
                             continue;
                         }
+                        Map<String, List<SampleMetadata>> patientSamplesMap = new HashMap<>();
+
                         // generate cmo labels for each sample metadata
                         List<SampleMetadata> updatedSamples = new ArrayList<>();
-                        for (SampleMetadata sample : request.getSamples()) {
-                            // generate cmo label
-                            String sampleCmoLabel = cmoLabelGeneratorService.generateCmoSampleLabel(sample);
+                        for (SampleMetadata sample : samples) {
+                            // get or request existing patient samples and update patient sample mapping
+                            if (!patientSamplesMap.containsKey(sample.getCmoPatientId())) {
+                                Message reply = messagingGateway.request(PATIENT_SAMPLES_REQUEST_TOPIC,
+                                        sample.getCmoPatientId());
+                                String replyData = mapper.readValue(
+                                        new String(reply.getData(), StandardCharsets.UTF_8),
+                                        String.class);
+                                patientSamplesMap.put(sample.getCmoPatientId(),
+                                        Arrays.asList(mapper.convertValue(replyData,
+                                                SampleMetadata[].class)));
+                            }
+                            List<SampleMetadata> existingSamples =
+                                    patientSamplesMap.get(sample.getCmoPatientId());
+
+                            // TODO resolve any issues that arise with errors in generating cmo label
+                            String sampleCmoLabel = cmoLabelGeneratorService.generateCmoSampleLabel(sample,
```

Comment:
If it wasn't inefficient, it would be cleaner to have the generator service get existing samples given a patient id.

Reply:
I thought about that but I wasn't sure about having a messaging gateway also in the label generator service as well as the message handler. Can discuss this further also